File: catowseriOS/catowser/Utils/DefaultTabProvider.swift
//
//  DefaultTabProvider.swift
//  catowser
//
//  Created by Andrei Ermoshin on 01/02/2019.
//  Copyright © 2019 Cotton/Catowser Andrei Ermoshin. All rights reserved.
//

import UIKit
import CoreBrowser
import FeaturesFlagsKit
import CottonBase

/// Provides default tab related constants, no need to be global actor, because read-only
@globalActor
final class DefaultTabProvider {
    static let shared = StateHolder()
    
    actor StateHolder: TabsStates {
        @MainActor var selected: Bool {
            UIDevice.current.userInterfaceIdiom == .pad
        }

        let blockPopups: Bool = false

        func topSites(_ isJsEnabled: Bool) -> [Site] {
            let array: [Site?]
            let settings: Site.Settings = .init(isPrivate: false,
                                                blockPopups: blockPopups,
                                                isJSEnabled: isJsEnabled,
                                                canLoadPlugins: true)
            let opennet = Site("https://opennet.ru", "OpenNet", settings)
            let yahooFinance = Site("https://finance.yahoo.com", "Yahoo Finance", settings)
            let github = Site("https://github.com", "GitHub", settings)
            #if DEBUG
            let mailToLink = Site("https://www.k8oms.net/links/mailto-link", "k8oms", settings)
            let mapsV1 = Site("https://www.apple.com/maps/", "apple maps", settings)
            array = [opennet, yahooFinance, github, mailToLink, mapsV1]
            #else
            array = [opennet, yahooFinance, github]
            #endif
            return array.compactMap {$0}
        }
        
        var addPosition: AddedTabPosition {
            get async {
                await FeatureManager.shared.tabAddPositionValue()
            }
        }

        var contentState: CoreBrowser.Tab.ContentType {
            get async {
                await FeatureManager.shared.tabDefaultContentValue()
            }
        }

        nonisolated var addSpeed: TabAddSpeed { .after(.milliseconds(300)) }

        nonisolated var defaultSelectedTabId: UUID { .notPossibleId }
    }
}

private extension UUID {
    static let notPossibleId: UUID = .init(uuid: (0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1))
}

File: catowserAndroid/app/src/main/java/org/cotton/app/utils/DefaultTabProvider.kt
package org.cotton.app.utils

import org.cotton.browser.content.data.tab.AddedTabPosition
import org.cotton.browser.content.data.tab.ContentType
import org.cotton.browser.content.data.tab.TabAddSpeed
import org.cotton.browser.content.service.tab.TabsStates
import java.util.UUID

class DefaultTabProvider : TabsStates {
    override val addPosition: AddedTabPosition = AddedTabPosition.AFTER_SELECTED
    override val contentState: ContentType = ContentType.TopSites
    override val addSpeed: TabAddSpeed = TabAddSpeed.Immediately
    override val defaultSelectedTabId: UUID = DefaultTab.notPossibleId
}

object DefaultTab {
    val notPossibleId: UUID by lazy {
        val bytes = ByteArray(4) {
            return@ByteArray when (it) {
                0 -> 0
                1 -> 255.toByte()
                2 -> 0
                3 -> 255.toByte()
                else -> 0
            }
        }
        UUID.nameUUIDFromBytes(bytes)
    }
}
